akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -439,25 +439,23 @@ final class FoldAsync[In, Out](zero: Out, f: (Out, In) ⇒ Future[Out]) extends
   def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) with InHandler with OutHandler {
     val decider = inheritedAttributes.get[SupervisionStrategy].map(_.decider).getOrElse(Supervision.stoppingDecider)
 
-    private var inFlight = 0
-
     private var aggregator: Out = zero
-    private var aggregating: Future[Out] = Future.successful(aggregator)
+    private var aggregated: Promise[Out] = Promise.successful(aggregator)
```

Comment:
Ah, to have separate callbacks. I think a solution with just the returned future, and a single callback that handle the case where in being closed when the future completes will be much cleaner (and avoid the extra allocation of a promise for each element).